clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -200,6 +200,99 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
+bool isZeroSizedArray(QualType Ty) {
+  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
+    return CATy->getSize() == 0;
+  return false;
+}
+
+Sema::DeviceDiagBuilder emitDeferredDiagnosticAndNote(Sema &S, SourceRange Loc,
+                                                      unsigned DiagID,
+                                                      SourceRange UsedAtLoc) {
+  Sema::DeviceDiagBuilder builder =
+      S.SYCLDiagIfDeviceCode(Loc.getBegin(), DiagID);
+  if (UsedAtLoc.isValid())
+    S.SYCLDiagIfDeviceCode(UsedAtLoc.getBegin(), diag::note_sycl_used_here);
+  return builder;
+}
+
+void checkSYCLVarType(Sema &S, QualType Ty, SourceRange Loc,
+                      llvm::DenseSet<QualType> Visited,
+                      SourceRange UsedAtLoc = SourceRange()) {
+  // not all variable types are supported in kernel contexts
+  // for any potentially unsupported types we issue a deferred diagnostic
+  // pass in the UsedAtLoc if a different location is needed to alert user to
+  // usage in SYCL context (example: struct member usage vs. declaration)
+
+  // zero length arrays
+  if (isZeroSizedArray(Ty))
+    emitDeferredDiagnosticAndNote(S, Loc, diag::err_typecheck_zero_array_size,
+                                  UsedAtLoc);
+
+  // sub-reference
+  while (Ty->isAnyPointerType() || Ty->isArrayType())
+    Ty = QualType{Ty->getPointeeOrArrayElementType(), 0};
+
+  // check types
+  // __int128, __int128_t, __uint128_t
+  if (Ty->isSpecificBuiltinType(BuiltinType::Int128) ||
```

Comment:
It looks like this and the __float128 check can be combined.

Reply:
they have slightly different conditionals. The float also checks for `hasFloat128Type`

Reply:
Yep, I saw that.  if (Ty->isSpecificBuiltinType(...) || Ty->isSpecificBuiltinType(...) || (Ty->isSpecificBuiltinType(...) && hasFloat128Type())

Reply:
That makes for a gnarly condition and if someone tampers with it in the future, it'll be easy to introduce errors. I'd prefer to keep the two cases separate. Would that be ok?

Reply:
I see it as the reverse... it makes it really easy to add additional types as we decide to prohibit them.  We have this exact pattern in the compiler all over the place, so it doesn't seem particularly gnarly to me.

Reply:
>  it makes it really easy to add additional types as we decide to prohibit them
Hmm.  I know of more type proscriptions that I am planning for future PRs, but none of them are simple.  In the future, as the type support in SPIRV grows, we are more likely going to need to remove some of these checks, narrow the scope of their conditions, and/or refine the diagnostic messaging shown the user.  For these reasons I think it'd be best to leave the cases separate as they are now.

But I'll defer to whatever you want to do, In all likelihood our discussion here is already more time spent on it than is warranted.    Let me know.

Reply:
My biggest concern is that despite the bodies of these two conditionals being the same, you have already implemented them differently. If the patch author can't get that right, I don't have hope for future revisions.

Reply:
I'd prefer them combined, but since you're probably not going to get ot this again this evening, we can let anyone else who has an opinion chime in overnight :)

Reply:
I prefer being aligned with the existing community code, so let's combine them if it can help.

Reply:
done